evolver/device.py

```diff
@@ -1,81 +1,54 @@
+from collections import defaultdict
 import time
-import pydantic
-from typing import Annotated
-from evolver.hardware.interface import SensorDriver, EffectorDriver
+
+from evolver.base import BaseConfig, BaseInterface, ConfigDescriptor
+from evolver.connection.interface import Connection
+from evolver.controller.interface import Controller
+from evolver.hardware.interface import EffectorDriver, HardwareDriver, SensorDriver
 from evolver.serial import EvolverSerialUART
+from evolver.settings import settings
 from evolver.history import HistoryServer
 
 
 DEFAULT_SERIAL = EvolverSerialUART
 DEFAULT_HISTORY = HistoryServer
-# pydantics import string alone does not generate a schema, which breaks openapi
-# docs. We wrap it to set schema explicitly.
-ImportString = Annotated[
-    pydantic.ImportString, pydantic.WithJsonSchema({'type': 'string', 'description': 'fully qualified class name'})
-]
-
-
-class ControllerDescriptor(pydantic.BaseModel):
-    driver: ImportString
-    config: dict = {}
-
-    def driver_from_descriptor(self, evolver):
-        conf = self.driver.Config.model_validate(self.config)
-        return self.driver(evolver, conf)
-
-
-class HardwareDriverDescriptor(ControllerDescriptor):
-    calibrator: ControllerDescriptor = None
-
-
-class EvolverConfig(pydantic.BaseModel):
-    vials: list = list(range(16))
-    hardware: dict[str, HardwareDriverDescriptor] = {}
-    controllers: list[ControllerDescriptor] = []
-    serial: ControllerDescriptor = ControllerDescriptor(driver=DEFAULT_SERIAL)
-    history: ControllerDescriptor = ControllerDescriptor(driver=DEFAULT_HISTORY)
-    enable_control: bool = True
-    enable_commit: bool = True
-    interval: int = 20
-
-
-class Evolver:
-    def __init__(self, config: EvolverConfig = EvolverConfig()):
-        self.hardware = {}
-        self.last_read = {}
-        self.controllers = []
-        self.update_config(config)
-
-    def update_config(self, config: EvolverConfig):
```

Comment:
As discussed, new instances will get created instead of reconfiguring an existing instance.